client/interface.go

```diff
@@ -0,0 +1,33 @@
+package client
+
+import (
+	"encoding"
+
+	"github.com/libp2p/go-libp2p-core/event"
+)
+
+// Client represents the drand Client interface.
+type Client interface {
+	// TextMarshaler (the 'MarshalText' method) allows serialization of the client
+	// state for subsequent restoration. The marshalled value can be passed to `New`
+	// to restore Client state.
+	encoding.TextMarshaler
+
+	// Get returns a the randomness at `round` or an error over a result channel.
+	// One result will be returned, with exactly one of Err or Data populated.
+	func Get(ctx context.Context, round uint64) <-chan Result
+
+	// RoundAt will return the most recent round of randomness that will be available
+	// at time for the current client.
+	func RoundAt(time time.Time) uint64
+}
+
+// Result represents the randomness for a single drand round.
+type Result struct {
+	Err error
+	Round uint64
+	Data []byte
+}
+
+// New creates a new Client, or a non-nill error if the input state cannot be parsed.
+func New(state []byte) (Client, error)
```

Comment:
I would like the interface for consuming randomness values to be distinct from the client implementation. So within a consuming process, most code just uses the `Client` interface, with `Get` and `RoundAt` etc, but somewhere close to `main` the caller constructs a concrete implementation like `NewGRPCClient`, `NewLibp2pClient`, `NewFakeClient`, etc. I do not think that hiding the concrete type of the thing being constructed here is helpful.

This would mean that `New` doesn't belong here (and generally could have a different type signature for different concrete implementations), and also that the `encoding.TextMarshaler` doesn't go in the `Client` interface, but perhaps in the `GRPCClient` interface/struct. The use of text marshalling for state construction should be left to the concrete implementation.

As an aside, I'm not sure where the motivation for that marshalling comes from.

Reply:
My interpretation here is that you expect to have your own protocol for learning where to learn random values from. Does that mean you'll take care of learning / updating Drand relays, and that's not something you expect the drand system to manage internally?

e.g. if the URL or bootstrap nodes a which a drand producer is making their shares available changes, you don't expect the drand client to be able to follow / learn, and will supply that through an out-of-band config option passed into a client node.

Reply:
That was my prior expectation, yes. If such functionality is included in some client implementation, I think it would still be specific to the concrete impl type rather than be on this interface.

For a libp2p-backed impl, I'm expecting that gossipsub would take care of that for us. For a direct point-to-point client, providing a list of relay addresses seems fine. (Possibly mutable at runtime, but again that's on the concrete impl, not this interface).

Reply:
I agree with @anorth here that there might be mulitple ways to fetch randomness and it's left to the user to choose which client to use, so we should have the `Client` be an interface and have specific instantiation of that interface, as what's alreay the case for HTTP and gRPC.

That being said for the "Marshaller" side, we should define a concrete  struct that defines all the public necessary information to get randomness (a list of relay nodes etc ) that these clients take into their constructor (+ other specific implementation details if necessary).

Reply:
There are two types of configs it seems like:
* static, implementation specific config (URL to poll from for an HTTP/GRPC-client. Libp2p keys / libp2p host to use for a gossipsub-client).
* current state - (this is a motivation for marshal/restore) - What is the most recent round that has been seen? what are the recently gossiped messages i've seen? This is probably most relevant to a gossip style client, but i wonder if it's worth having a single interface to express the desire to have a clear interface for checkpoint / restore lifecycle

Reply:
Right now the PR doesn't seem to include these information "last round seen" - I wouldn't bother adding an interface constraint to something not implemented / used for now.

Reply:
Providing a auto-configuring `New` is removed from the interface. i believe this is now addressed.